### ISSUE TITLE
ARM: auto-register for the requested RP

### ISSUE THREAD

**Issue by ahmetalpbalkan** [Azure/azure-xplat-cli#2741]
CLI Version: **0.9.19**
OS Type: Mac
Installation via:  npm

Mode: **ARM**

For a first-time user trying to do something, telling them an error about failure to register to a namespace is confusing-AS-HELL. If I'd be a first time user:

- what the hell is a namespace
- so why did not the [docs](https://azure.microsoft.com/en-us/documentation/articles/key-vault-manage-with-cli/) mention that
- how do I register to a namespace

We should either register for RPs automatically on user’s behalf (why not) or (for whatever reason why RP registration exists in the first place) detect the missing registration early on and fail with a message that says "this is how you should register for a namespace".

```
$ azure keyvault create --vault-name acr-secrets -g acr-0 -l westus
info:    Executing command keyvault create
+ Checking pre-condition
warn:    No object ID is selected. The current user's object ID will be used by default
+ Creating vault acr-secrets
error:   The subscription is not registered to use namespace 'Microsoft.KeyVault'.
info:    Error information has been recorded to /Users/alp/.azure/azure.err
error:   keyvault create command failed
```

**Comment by yugangw-msft** [comment]
@amarzavery @msfcolombo, we do have [clear logic](https://github.com/Azure/azure-xplat-cli/blob/dev/lib/util/utils.js#L527) to tell user what to do. It is possible the error detecting does not work out or keyvalult doesn't go through the common client creation process

**Comment by amarzavery** [comment]
Yes. Looks like keyvault is taking another path for instantiating their client.

**Comment by msfcolombo** [comment]
@amarzavery @yugangw-msft Key Vault is using createResourceClient which is shared at least by `webapp`, `vm` and `network` sub-commands as you can see [here](https://github.com/Azure/azure-xplat-cli/search?utf8=%E2%9C%93&q=createResourceClient). So either other sub-commands are affected, or the logic is buggy. Can you confirm which is happening?

**Comment by squillace** [comment]
If I may, from the user's perspective, this should ALWAYS automatically register IF the specific provider doesn't cost money or require some specific "state-changing" feature. Azure is a unified platform that just HAPPENS to be implemented by asynchronous providers -- but we do not teach naive users that they must modify our underlying implementation. If they try to use keyvault, keyvault MUST respond "yes" in any case in which it will not automatically cost money. 

If, on the other hand, someone chooses to dive in to unregister, then they're off the reservation, and we should automatically reregister, because there is no known scenario in which a user turns "off" a feature that costs them nothing. If they wish to restrict its use, they have roles and policies to do so at the AAD level.

This is separate from the underlying technical discussion. There should never, ever be a good argument for forcing the customer to understand something _that has absolutely no effect on them but adding a lovely bonus feature of Azure_. If there is, I'd love to hear it, but I haven't yet. (My guess is that other providers that use this code are already automatically registered, thus the issue doesn't arise.... but I could be wrong there.)

**Comment by squillace** [comment]
Also: Do our competitors force the user to do this? That might be a good question, either way.

**Comment by msfcolombo** [comment]
@amarzavery @yugangw-msft We checked Azure Powershell and their shared code is auto-registering providers (see [here](https://github.com/Azure/azure-powershell/blob/dev/src/ResourceManager/Common/Commands.ResourceManager.Common/RPRegistrationDelegatingHandler.cs#L64)). This is a lack of symmetry: When the person uses PSH to create a Key Vault (or any other ARM resource), he/she doesn't have to register explicitly. When using CLI, explicit registration is required. We need an authoritative answer on this.

**Comment by iboware** [comment]
Same thing happens to me too when I try to create a resource through azure-cli on Linux with a brand new subscription. If I create and delete a resource of same type from Azure Portal then cli will start to work. It is very strange and confusing.
